optional/slides/slides.md

```diff
@@ -0,0 +1,207 @@
+
+---
+
+## Introducing the Optional class, for Java 6
+
+### Yes! Guava has `com.google.common.base.Optional<T>` too!
+
+Unlike the Java one this implements _Serializable_!
+
+Has similar methods but with different names:
+* `fromNullable` and `ofNullable`
+* `absent` and `empty`
+* `or` and `orElse`
+* `transform` and `map`/`flatMap`
+
+But sometimes not: `get`, `isPresent`.
+
+???
+This can be used like a field.
+
+---
+
+## Example: Guava
+
+```
+public class Insurance {
+    private String name;
+    public String getName() { return name; }
+}
+
+public String getInsuranceName(Insurance insurance) {
+	return Optional.fromNullable(insurance)
+		.transform(new Function<Insurance, String>() {
+			@Override public String apply(Insurance input) {
+				return input.getName();
+			}
+    }).or("Unknow");
+}
+```
+
+Java 6 style! Still ugly...
+
+---
+
+## Example: Method returns Optional&lt;T&gt;
+
+Returning Optional adds explicitly the possibility that there might not be a name for that given Insurance.
+
+This means that the caller of the method is explicitly forced by the type system to think about and deal with the possibility that there might not be a name with that Insurance.
+
+```
+public class Insurance {
+    private String name;
+    public Optional<String> getName() {
+    	return Optional.fromNullable(name);
+    }
+}
+```
+
+---
+
+## Example: Java 8
+
+```
+public class Insurance {
+	private String name;
+	public Optional<String> getName() {
+		return Optional.ofNullable(name);
+	}
+}
+```
+```
+insurance.getName().orElse("Unknown");
+```
+
+Bye Bye NullPointerException!
+
+---
+
+## Interesting methods
+
+### Guava
+
+* `Set<T> asSet()`: immutable singleton Set if it is present; an empty immutable Set otherwise.
+* `T or(T)`: contained instance if it is present; defaultValue otherwise.
+* `Optional<V> transform(Function<? super T, V>)`: if the instance is present, it is transformed with the given Function; otherwise, Optional.absent is returned.
+
+---
+
+## Interesting methods
+
+### Java
+
+* `Optional<T> filter(Predicate<? super T>)`: when predicate is false return an empty optional
+* `Optional<U> map(Function<? super T, ? extends U>)`: if a value is present, apply the provided mapping function to it
+* `Optional<U> flatMap(Function<? super T, Optional<U>>)`: like map, but the function result in another optional
+* `void ifPresent(Consumer<? super T>)`: if a value is present, invoke the specified consumer with the value
+* `T orElse(T)`
+* `<X extends Throwable> T orElseThrow(Supplier<? extends X>)`
+
+---
+
+# What is Optional not trying to solve
+
+Optional is not meant to be a mechanism to avoid all types of null pointers. The mandatory input parameters of methods and constructors still have to be tested for example.
+
+## What is wrong with just returning null?
+
+The problem is that **the caller of the function might not have read the javadoc** for the method, and forget about handling the null case.
```

Comment:
the whole key is "ambiguity". It should be mentioned here or right at the start of the talk.